q2_demux/tests/test_filter.py

```diff
@@ -0,0 +1,183 @@
+        dir_files = os.listdir(str(test))
+        meta_ids = meta.get_ids(where=where)
+        dir_ids = [(lambda x: dir_files[x].split('_')[0::3])(x) for
+                   x in range(len(dir_files))]
+
+        for id in meta_ids:
+            self.assertIn([id, 'R1'], dir_ids)
+            self.assertIn([id, 'R2'], dir_ids)
+        self.assertTrue(len(dir_ids) == 2 * len(meta_ids))
+
+    def _paired_exclude(self, meta, where):
+        test = filter.filter_samples(self.sample_paired, meta, where, True)
+        dir_files = os.listdir(str(test))
+        meta_ids = meta.get_ids(where=where)
+        dir_ids = [(lambda x: dir_files[x].split('_')[0::3])(x) for
+                   x in range(len(dir_files))]
+        exp_ids = set(self.manifest_paired.index) - meta_ids
+
+        for id in exp_ids:
+            self.assertIn([id, 'R1'], dir_ids)
+            self.assertIn([id, 'R2'], dir_ids)
+        self.assertEqual(set((lambda x: dir_ids[x][0])(x) for x
+                         in range(len(dir_ids))), exp_ids)
+
+    def test_filter_paired(self):
+        metas = [self.meta_paired_all, self.meta_paired_subset]
+        wheres = [None, "Study='A'", "Study='B'", "Study='A' OR Study='B'",
+                  "Study='A' AND Study='B'"]
+        for meta in metas:
+            for where in wheres:
+                with self.subTest(meta=meta, where=where):
+                    self._paired_no_exclude(meta, where)
+
+    def test_filter_paired_exclude(self):
+        metas = [self.meta_paired_all, self.meta_paired_subset]
+        wheres = [None, "Study='A'", "Study='B'", "Study='A' OR Study='B'",
+                  "Study='A' AND Study='B'"]
+        for meta in metas:
+            for where in wheres:
+                with self.subTest(meta=meta, where=where):
+                    self._paired_exclude(meta, where)
+
+    def test_filter_paired_none_exclude(self):
+        wheres = [None, "Study='A'", "Study='B'", "Study='A' OR Study='B'",
+                  "Study='A' AND Study='B'"]
+        for where in wheres:
+            with self.subTest(where=where):
+                self._paired_exclude(self.meta_single_none, where)
+
+    def test_filter_paired_none_no_exclude_no_exception(self):
+        wheres = ["Study='B'", "Study='A' AND Study='B'"]
+        for where in wheres:
+            with self.subTest(where=where):
+                self._paired_no_exclude(self.meta_single_none, where)
+
+    def test_filter_paired_none_exception(self):
+        wheres = [None, "Study='A'", "Study='A' OR Study='B'"]
+        for where in wheres:
+            with self.subTest(where=where):
+                with self.assertRaisesRegex(KeyError, '1'):
+                    filter.filter_samples(self.sample_paired,
+                                          self.meta_paired_none, where,
+                                          False)
+
+
+if __name__ == '__main__':
+    unittest.main()
```

Comment:
These test cases are incredibly comprehensive, you did an excellent job with them. I think I have an idea which may assist in avoiding using the original implementation. But it might be easiest to describe in person.